devices/usbc-cdc/ctrl.c

```diff
@@ -116,49 +122,70 @@ static int ctrl_dtdInit(int endpt, int inQH, int outQh)
 }
 
 
-static int ctrl_initEndptQh(int endpt, int dir, endpt_data_t *endpt_init)
+static u32 ctrl_initEndptQh(int endpt, int dir, endpt_data_t *endpt_init)
 {
-	u32 setup = 0, tmp;
+	u32 setup, caps;
 	int qh = endpt * 2 + dir;
 
-	ctrl_allocBuff(endpt, dir);
+	if (ctrl_allocBuff(endpt, dir) < 0) {
+		return -ENOMEM;
+	}
 
-	tmp = endpt_init->caps[dir].max_pkt_len << 16;
-	tmp |= endpt_init->caps[dir].ios << 15;
-	tmp |= endpt_init->caps[dir].zlt << 29;
-	tmp |= endpt_init->caps[dir].mult << 30;
-	ctrl_common.dc->endptqh[qh].caps = tmp;
+	caps = (endpt_init->caps[dir].max_pkt_len & 0x7ff) << 16;
+	caps |= (endpt_init->caps[dir].ios & 1) << 15;
+	caps |= (endpt_init->caps[dir].zlt & 1) << 29;
+	caps |= (endpt_init->caps[dir].mult & 3) << 30;
+	ctrl_common.dc->endptqh[qh].caps = caps;
 	ctrl_common.dc->endptqh[qh].dtd_next = 1;
 
-	setup |= endpt_init->ctrl[dir].data_toggle << (6 + dir * 16);
-	setup |= endpt_init->ctrl[dir].type << (2 + dir * 16);
+	setup = (endpt_init->ctrl[dir].data_toggle & 1) << 6;
+	setup |= (endpt_init->ctrl[dir].type & 3) << 2;
+	setup |= (endpt_init->ctrl[dir].stall & 1);
 
-	return setup;
+	return dir != 0 ? setup << 16 : setup;
+}
+
+
+void ctrl_initQtd(void)
+{
+	/*
+	 * Reset qtd offset for non control endpoints every time device is re/connected to
+	 * host, at any time when desc_setup(REQ_SET_ADDRESS) is initiated
+	 */
+
+	ctrl_common.qtdOffs = 0;
 }
 
 
 int ctrl_endptInit(int endpt, endpt_data_t *endpt_init)
 {
-	s8 i;
+	u8 i;
 	int res = EOK;
 	u32 setup = 0;
 
-	if (endpt == 0)
-		return -1;
+	if (endpt <= 0) {
+		return -ENXIO;
+	}
 
-	if ((res = ctrl_dtdInit(endpt, endpt_init->caps[USB_ENDPT_DIR_IN].init, endpt_init->caps[USB_ENDPT_DIR_OUT].init)) != EOK)
+	res = ctrl_dtdInit(endpt, endpt_init->caps[USB_ENDPT_DIR_IN].init, endpt_init->caps[USB_ENDPT_DIR_OUT].init);
+	if (res != EOK) {
 		return res;
+	}
 
+	/* Setup RX & TX endpoints */
 	for (i = 0; i < ENDPOINTS_DIR_NB; ++i) {
-		if (endpt_init->caps[i].init)
+		if (endpt_init->caps[i].init != 0) {
 			setup |= ctrl_initEndptQh(endpt, i, endpt_init);
+		}
 	}
 
 	*(ctrl_common.dc->base + endptctrl0 + endpt) = setup;
 
+	/* Enable endpoints */
 	for (i = 0; i < ENDPOINTS_DIR_NB; ++i) {
-		if (endpt_init->caps[i].init)
-			*(ctrl_common.dc->base + endptctrl0 + endpt) |= 1 << (7 + i * 16);
+		if (endpt_init->caps[i].init != 0) {
+			*(ctrl_common.dc->base + endptctrl0 + endpt) |= 1u << (7 + i * 16);
```

Comment:
1U is preferred to 1u because you can't do 1ul or 1ull (as lowecase L can be mistaken for 1), but 1UL and 1ULL is ok.

Reply:
Yup.